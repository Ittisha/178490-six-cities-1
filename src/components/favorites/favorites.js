import * as React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {NavLink} from 'react-router-dom';

import {history} from '../../history';
import {CityLink} from '../city-link/city-link';
import {Card} from '../card/card';
import {Operation} from '../../reducers/favorites/favorites';
import {getFavorites} from '../../reducers/selectors';
import {ActionCreator as CitiesActionCreator} from '../../reducers/cities/cities';

export class Favorites extends React.PureComponent {
  constructor(props) {
    super(props);

    this._handleLinkClick = this._handleLinkClick.bind(this);
  }
  componentDidMount() {
    const {loadFavorites} = this.props;
    loadFavorites();
  }

  render() {
    const {favorites} = this.props;
    const hasFavorites = favorites && Object.keys(favorites).length !== 0;
    const emptyClass = !hasFavorites ? `page__main--favorites-empty` : ``;

    return (<React.Fragment>
      <main className={`page__main page__main--favorites ${emptyClass}`}>
        <div className="page__favorites-container container">
          {!hasFavorites && (
            <section className="favorites favorites--empty">
              <h1 className="visually-hidden">Favorites (empty)</h1>
              <div className="favorites__status-wrapper">
                <b className="favorites__status">Nothing yet saved.</b>
                <p className="favorites__status-description">Save properties to narrow down search or plan yor future trips.</p>
              </div>
            </section>
          )}
          {hasFavorites && (
            <section className="favorites">
              <h1 className="favorites__title">Saved listing</h1>
              <ul className="favorites__list">
                {Object.entries(favorites).map(([city, offers]) => <li
                  key={`favorite-${city}`}
                  className="favorites__locations-items"
                >
                  <div className="favorites__locations locations locations--current">
                    <CityLink
                      city={city}
                      cityCoords={offers[0].cityCoords}
                      cityZoom={offers[0].cityZoom}
                      isActive={true}
                      isDiv={true}
                      onLinkClick={this._handleLinkClick}
                    />
                  </div>
                  <div className="favorites__places">
                    {offers.map((offer) => <Card
                      key={`${city}-${offer.id}`}
                      cardClass="favorites__card"
                      imageWrapperClass="favorites__image-wrapper"
                      infoClass="favorites__card-info"
                      apartment={offer}
                      isSmall={true}
                    />)}
                  </div>
                </li>)}
              </ul>
            </section>
          )}
        </div>
      </main>
      <footer className="footer container">
        <NavLink
          to={`/`}
          className="footer__logo-link"
        >
          <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33" />
        </NavLink>
      </footer>
    </React.Fragment>
    );
  }

  _handleLinkClick(cityToChange) {
    const {changeCity} = this.props;
    changeCity(cityToChange);
    history.push(`/`);
  }
}

Favorites.propTypes = {
  loadFavorites: PropTypes.func.isRequired,
  favorites: PropTypes.object.isRequired,
  changeCity: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  favorites: getFavorites(state),
});

const mapDispatchToProps = (dispatch) => ({
  loadFavorites: () => dispatch(Operation.loadFavorites()),
  changeCity: (city) => dispatch(CitiesActionCreator.changeCity(city)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
