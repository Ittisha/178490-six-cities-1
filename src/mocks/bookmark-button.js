export const MOCK_PROPS = {
  offerId: 1,
  isInBookmarks: false,
  className: ``,
  width: 15,
  height: 12,
};
