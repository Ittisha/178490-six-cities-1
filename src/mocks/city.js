export const INIT_CITY = {
  name: `Amsterdam`,
  coords: [52.38333, 4.9],
  zoom: 12,
};

export const MOCK_CITIES = [`Amsterdam`, `Paris`];

export const MOCK_CITIES_COORDS = {
  Amsterdam: [52.38333, 4.9],
  Paris: [48.864716, 2.349014],
};

export const MOCK_CITIES_ZOOM = {
  Amsterdam: 17,
  Paris: 15,
};
