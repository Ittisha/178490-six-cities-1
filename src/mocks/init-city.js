export const INIT_CITY = {
  name: `Amsterdam`,
  coords: [52.38333, 4.9],
};
