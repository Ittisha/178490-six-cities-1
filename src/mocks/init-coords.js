export const INIT_COORDS = [52.38333, 4.9];
