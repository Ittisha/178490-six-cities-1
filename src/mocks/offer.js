export const OFFER = {
  "city": {
    "name": `Cologne`,
    "location": {
      "latitude": 50.938361,
      "longitude": 6.959974,
      "zoom": 13}},
  "preview_image": `https://es31-server.appspot.com/six-cities/static/hotel/17.jpg`,
  "images": [
    `https://es31-server.appspot.com/six-cities/static/hotel/20.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/2.jpg`,
  ],
  "title": `The house among olive `,
  "is_favorite": false,
  "is_premium": false,
  "rating": 2.5,
  "type": `house`,
  "bedrooms": 2,
  "max_adults": 6,
  "price": 757,
  "goods": [
    `Coffee machine`,
    `Baby seat`,
    `Dishwasher`,
    `Fridge`,
    `Washing machine`,
    `Breakfast`,
  ],
  "host": {
    "id": 25,
    "name": `Angelina`,
    "is_pro": true,
    "avatar_url": `img/avatar-angelina.jpg`
  },
  "description": `This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.`,
  "location": {
    "latitude": 50.945361,
    "longitude": 6.935974,
    "zoom": 16
  },
  "id": 1
};
