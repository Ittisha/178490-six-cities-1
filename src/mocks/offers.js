export const OFFERS = [
  {
    isPremium: true,
    isInBookmarks: false,
    title: `Beautiful & luxurious apartment at great location`,
    photoUrl: `img/apartment-01.jpg`,
    price: 120,
    type: `Apartment`,
    rating: `93%`,
    id: 1,
    coordinates: [52.3909553943508, 4.85309666406198],
  },
  {
    isPremium: false,
    isInBookmarks: true,
    title: `Wood and stone place`,
    photoUrl: `img/room.jpg`,
    price: 80,
    type: `Private room`,
    rating: `80%`,
    id: 2,
    coordinates: [52.369553943508, 4.85309666406198],
  },
  {
    isPremium: false,
    isInBookmarks: false,
    title: `Canal View Prinsengracht`,
    photoUrl: `img/apartment-02.jpg`,
    price: 132,
    type: `Apartment`,
    rating: `80%`,
    id: 3,
    coordinates: [52.3909553943508, 4.929309666406198],
  },
  {
    isPremium: false,
    isInBookmarks: false,
    title: `Nice, cozy, warm big bed apartment`,
    photoUrl: `img/apartment-03.jpg`,
    price: 180,
    type: `Apartment`,
    rating: `80%`,
    id: 4,
    coordinates: [52.3809553943508, 4.939309666406198],
  }
];
