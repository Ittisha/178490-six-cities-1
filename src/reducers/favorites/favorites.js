import {toast} from 'react-toastify';
import {mapOffers} from '../../mappers/map-offers';
import {ActionCreator as OffersActionCreator} from '../offers/offers-data';
import {SUCCESS_STATUS} from '../../consts';

const getBookmarkAction = (isInBookmarks) => isInBookmarks ? `added to` : `removed from`;

export const ActionType = {
  UPDATE_FAVORITES: `UPDATE_FAVORITES`,
};

export const ActionCreator = {
  updateFavorites: (offer) => ({
    type: ActionType.UPDATE_FAVORITES,
    payload: offer,
  })
};

export const Operation = {
  updateFavorite: (isInBookmarks, id) => (dispatch, _getState, api) => api.post(`/favorite/${id}/${isInBookmarks}`)
    .then((response) => {
      if (response.status === SUCCESS_STATUS) {
        const data = mapOffers([response.data])[0];
        dispatch(ActionCreator.updateFavorites(data));
        dispatch(OffersActionCreator.updateOffers(data));
        toast.success(`The offer was successfully ${getBookmarkAction(isInBookmarks)} favorites`);
        return;
      }
      toast.error(response.response.data.error);
    })
    .catch((error) => {
      toast.error(error.message);
    }),
};

