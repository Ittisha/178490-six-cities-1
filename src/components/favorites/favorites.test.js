import * as React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import {Favorites} from './favorites';
import {MOCK_FAVORITES} from '../../mocks/offers';

it(`Favorites renders correctly`, () => {
  const renderer = new ShallowRenderer();
  const tree = renderer
    .render(
        <Favorites
          loadFavorites={() => {}}
          favorites={MOCK_FAVORITES}
          changeCity={() => {}}
        />
    );

  expect(tree).toMatchSnapshot();
});
