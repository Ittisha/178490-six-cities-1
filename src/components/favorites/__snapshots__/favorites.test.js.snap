// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites renders correctly 1`] = `
<React.Fragment>
  <main
    className="page__main page__main--favorites "
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <CityLink
                city="Amsterdam"
                cityCoords={
                  Array [
                    52.38333,
                    4.9,
                  ]
                }
                cityZoom={15}
                isActive={true}
                isDiv={true}
                onLinkClick={[Function]}
              />
            </div>
            <div
              className="favorites__places"
            >
              <Card
                apartment={
                  Object {
                    "bedrooms": 2,
                    "cityCoords": Array [
                      52.38333,
                      4.9,
                    ],
                    "cityName": "Amsterdam",
                    "cityZoom": 15,
                    "coordinates": Array [
                      52.3909553943508,
                      4.929309666406198,
                    ],
                    "description": "description",
                    "goods": Array [
                      "Something",
                    ],
                    "host": Object {
                      "avatar": "apartment-01.jpg",
                      "id": 2,
                      "isPro": true,
                      "name": "Name",
                    },
                    "id": 3,
                    "images": Array [
                      "img/apartment-01.jpg",
                    ],
                    "isInBookmarks": false,
                    "isPremium": false,
                    "maxAdults": 4,
                    "photoUrl": "img/apartment-02.jpg",
                    "price": 132,
                    "rating": 4,
                    "title": "Canal View Prinsengracht",
                    "type": "Apartment",
                    "zoom": 16,
                  }
                }
                cardClass="favorites__card"
                imageWrapperClass="favorites__image-wrapper"
                infoClass="favorites__card-info"
                isSmall={true}
              />
              <Card
                apartment={
                  Object {
                    "bedrooms": 2,
                    "cityCoords": Array [
                      52.38333,
                      4.9,
                    ],
                    "cityName": "Amsterdam",
                    "cityZoom": 15,
                    "coordinates": Array [
                      52.3809553943508,
                      4.939309666406198,
                    ],
                    "description": "description",
                    "goods": Array [
                      "Something",
                    ],
                    "host": Object {
                      "avatar": "apartment-01.jpg",
                      "id": 2,
                      "isPro": true,
                      "name": "Name",
                    },
                    "id": 4,
                    "images": Array [
                      "img/apartment-01.jpg",
                    ],
                    "isInBookmarks": false,
                    "isPremium": false,
                    "maxAdults": 4,
                    "photoUrl": "img/apartment-03.jpg",
                    "price": 180,
                    "rating": 4,
                    "title": "Nice, cozy, warm big bed apartment",
                    "type": "Apartment",
                    "zoom": 16,
                  }
                }
                cardClass="favorites__card"
                imageWrapperClass="favorites__image-wrapper"
                infoClass="favorites__card-info"
                isSmall={true}
              />
            </div>
          </li>
        </ul>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <NavLink
      className="footer__logo-link"
      to="/"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </NavLink>
  </footer>
</React.Fragment>
`;
